galsim/phase_psf.py

```diff
@@ -0,0 +1,1399 @@
+
+    def __eq__(self, other):
+        return self._layers == other._layers
+
+    def __ne__(self, other): return not self == other
+
+    __hash__ = None  # Mutable means not hashable.
+
+    def _update_attrs(self):
+        # Update object attributes for current set of layers.  Currently the only attributes are
+        # self.time_step and self.rng.
+        # Could have made these each a @property instead of defining _update_attrs(), but then
+        # failures would occur late rather than early, which makes debugging more difficult.
+
+        # Each layer must have same value for time_step or no attr time_step.
+        time_step = set([l.time_step for l in self if hasattr(l, 'time_step')])
+        if len(time_step) == 0:
+            self.time_step = None
+        elif len(time_step) == 1:
+            self.time_step = time_step.pop()
+        else:
+            raise ValueError("Layer time steps must all be identical or None")
+
+        # If any of the wrapped PhaseScreens have an rng, then eval(repr(screen_list)) will run, but
+        # fail to round-trip to the original object.  So we search for that here and set/delete a
+        # dummy rng sentinel attribute so do_pickle() will know to skip the obj == eval(repr(obj))
+        # test.
+        self.__dict__.pop('rng', None)
+        if any(hasattr(l, 'rng') for l in self):
+            self.rng = None
+
+    def advance(self):
+        """Advance each phase screen in list by self.time_step."""
+        for layer in self:
+            try:
+                layer.advance()
+            except AttributeError:
+                # Time indep phase screen.
+                pass
+
+    def advance_by(self, dt):
+        """Advance each phase screen in list by specified amount of time.
+
+        @param dt  Amount of time in seconds by which to update the screens.
+        @returns   The actual amount of time updated, which can potentially (though not necessarily)
+                   differ from `dt` when `dt` is not a multiple of self.time_step.
+        """
+        for layer in self:
+            try:
+                out = layer.advance_by(dt)
+            except AttributeError:
+                # Time indep phase screen
+                pass
+        return out
+
+    def reset(self):
+        """Reset phase screens back to time=0."""
+        for layer in self:
+            try:
+                layer.reset()
+            except AttributeError:
+                # Time indep phase screen
+                pass
+
+    def wavefront(self, *args, **kwargs):
+        """ Compute wavefront due to phase screen.
+
+        Wavefront here indicates the distance by which the physical wavefront lags or leads the
+        ideal plane wave (pre-optics) or spherical wave (post-optics).
+
+        @param aper     `galsim.Aperture` over which to compute wavefront.
+        @param theta_x  x-component of field angle corresponding to center of output array.
```

Comment:
It might be less commonly used than `makePSF`, but I think we should probably keep this routine user accessible.  I fleshed-out the docstring here in PhaseScreenList and in the wavefront() methods in phase_screens.py.  Here's the new version for PhaseScreenList:

```
        """ Compute cumulative wavefront due to all phase screens in PhaseScreenList.

        Wavefront here indicates the distance by which the physical wavefront lags or leads the
        ideal plane wave (pre-optics) or spherical wave (post-optics).

        @param aper     `galsim.Aperture` over which to compute wavefront.
        @param theta_x  x-component of field angle corresponding to center of output array, as a
                        galsim.Angle instance.  [default: 0*galsim.degrees]
        @param theta_y  y-component of field angle corresponding to center of output array, as a
                        galsim.Angle instance.  [default: 0*galsim.degrees]
        @param compact  If true, then only return wavefront for illuminated pixels in a
                        single-dimensional array congruent with array[aper.illuminated].  Otherwise,
                        return wavefront as a 2d array for the full Aperture pupil plane.
                        [default: True]
        @returns        Wavefront lag or lead in nanometers over aperture.
        """
```